lib/Frenetic_NetKAT_Compiler.ml

```diff
@@ -526,17 +541,21 @@ module Pol = struct
     | Seq (p,q) -> Seq (of_pol ing p, of_pol ing q)
     | Star p -> Star (of_pol ing p)
     | Link (s1,p1,s2,p2) ->
-      (* SJS: This is not the true sematnics of a link! This is a hack that works for now,
-         but we will need to use the correct encoding once we start doing things like global
-         optimization or deciding equivalence. *)
       let post_link = match ing with
         | None -> filter_loc s2 p2
         | Some ing ->
-            Frenetic_NetKAT_Optimize.(mk_and (Test (Switch s2)) (mk_not ing))
-            |> mk_filter
-      in
-      mk_big_seq [filter_loc s1 p1; Dup; post_link ]
-    | VLink _ -> assert false (* SJS / JNF *)
+          Frenetic_NetKAT_Optimize.(mk_and (Test (Switch s2)) (mk_not ing))
+          |> mk_filter in
+      let link = mk_seq (mk_mod (Switch s2)) (mk_mod (Location (Physical p2))) in
+      mk_big_seq [filter_loc s1 p1; link; Dup; post_link]
```

Comment:
This change brakes the global compiler. You're conflating the parts of the program that need to be implemented by the network topology (`link`) with the parts that need to be implemented by the switches (`filter_loc s1 p2`, `post_link`, etc...). But when we extract a local program, we want only the switch stuff, not the topology stuff. You're erasing some of the topology stuff by deleting switch field modifications in `erase_switch_fields`, but you're not erasing the port field modifications.